File: AccuPay/Domain/Employees/Employee.vb
Option Strict On

Imports System.ComponentModel.DataAnnotations
Imports System.ComponentModel.DataAnnotations.Schema
Imports AccuPay.Data
Imports AccuPay.Enums
Imports PayrollSys

Namespace Global.AccuPay.Entity

    <Table("employee")>
    Public Class Employee
        Implements IEmployee

        <Key>
        Public Property RowID As Integer? Implements IEmployee.RowID

        Public Property Created As Date Implements IEmployee.Created
        Public Property CreatedBy As Integer? Implements IEmployee.CreatedBy
        Public Property LastUpd As Date? Implements IEmployee.LastUpd
        Public Property LastUpdBy As Integer? Implements IEmployee.LastUpdBy
        Public Property OrganizationID As Integer? Implements IEmployee.OrganizationID
        Public Property PositionID As Integer? Implements IEmployee.PositionID
        Public Property PayFrequencyID As Integer? Implements IEmployee.PayFrequencyID
        Public Property Salutation As String Implements IEmployee.Salutation
        Public Property FirstName As String Implements IEmployee.FirstName
        Public Property MiddleName As String Implements IEmployee.MiddleName
        Public Property LastName As String Implements IEmployee.LastName
        Public Property Surname As String Implements IEmployee.Surname

        <Column("EmployeeID")>
        Public Property EmployeeNo As String Implements IEmployee.EmployeeNo

        Public Property TinNo As String Implements IEmployee.TinNo
        Public Property SssNo As String Implements IEmployee.SssNo
        Public Property HdmfNo As String Implements IEmployee.HdmfNo
        Public Property PhilHealthNo As String Implements IEmployee.PhilHealthNo
        Public Property EmploymentStatus As String Implements IEmployee.EmploymentStatus
        Public Property EmailAddress As String Implements IEmployee.EmailAddress
        Public Property WorkPhone As String Implements IEmployee.WorkPhone
        Public Property HomePhone As String Implements IEmployee.HomePhone
        Public Property MobilePhone As String Implements IEmployee.MobilePhone
        Public Property HomeAddress As String Implements IEmployee.HomeAddress
        Public Property Nickname As String Implements IEmployee.Nickname
        Public Property JobTitle As String Implements IEmployee.JobTitle
        Public Property Gender As String Implements IEmployee.Gender
        Public Property EmployeeType As String Implements IEmployee.EmployeeType
        Public Property MaritalStatus As String Implements IEmployee.MaritalStatus
        Public Property BirthDate As Date Implements IEmployee.BirthDate
        Public Property StartDate As Date Implements IEmployee.StartDate
        Public Property TerminationDate As Date? Implements IEmployee.TerminationDate
        Public Property NoOfDependents As Integer? Implements IEmployee.NoOfDependents
        Public Property UndertimeOverride As Boolean Implements IEmployee.UndertimeOverride
        Public Property OvertimeOverride As Boolean Implements IEmployee.OvertimeOverride
        Public Property NewEmployeeFlag As Boolean Implements IEmployee.NewEmployeeFlag
        Public Property LeaveBalance As Decimal Implements IEmployee.LeaveBalance
        Public Property SickLeaveBalance As Decimal Implements IEmployee.SickLeaveBalance
        Public Property MaternityLeaveBalance As Decimal Implements IEmployee.MaternityLeaveBalance
        Public Property OtherLeaveBalance As Decimal Implements IEmployee.OtherLeaveBalance

        <Column("LeaveAllowance")>
        Public Property VacationLeaveAllowance As Decimal Implements IEmployee.LeaveAllowance

        Public Property SickLeaveAllowance As Decimal Implements IEmployee.SickLeaveAllowance
        Public Property MaternityLeaveAllowance As Decimal Implements IEmployee.MaternityLeaveAllowance
        Public Property OtherLeaveAllowance As Decimal Implements IEmployee.OtherLeaveAllowance
        Public Property LeavePerPayPeriod As Decimal Implements IEmployee.LeavePerPayPeriod
        Public Property SickLeavePerPayPeriod As Decimal Implements IEmployee.SickLeavePerPayPeriod
        Public Property MaternityLeavePerPayPeriod As Decimal Implements IEmployee.MaternityLeavePerPayPeriod
        Public Property OtherLeavePerPayPeriod As Decimal Implements IEmployee.OtherLeavePerPayPeriod
        Public Property AlphalistExempted As Boolean Implements IEmployee.AlphalistExempted
        Public Property WorkDaysPerYear As Decimal Implements IEmployee.WorkDaysPerYear
        Public Property DayOfRest As Integer? Implements IEmployee.DayOfRest
        Public Property AtmNo As String Implements IEmployee.AtmNo
        Public Property BankName As String Implements IEmployee.BankName
        Public Property CalcHoliday As Boolean Implements IEmployee.CalcHoliday
        Public Property CalcSpecialHoliday As Boolean Implements IEmployee.CalcSpecialHoliday
        Public Property CalcNightDiff As Boolean Implements IEmployee.CalcNightDiff
        Public Property CalcNightDiffOT As Boolean Implements IEmployee.CalcNightDiffOT
        Public Property CalcRestDay As Boolean Implements IEmployee.CalcRestDay
        Public Property CalcRestDayOT As Boolean Implements IEmployee.CalcRestDayOT
        Public Property DateRegularized As Date? Implements IEmployee.DateRegularized
        Public Property DateEvaluated As Date? Implements IEmployee.DateEvaluated
        Public Property RevealInPayroll As Boolean Implements IEmployee.RevealInPayroll
        Public Property LateGracePeriod As Decimal Implements IEmployee.LateGracePeriod
        Public Property OffsetBalance As Decimal Implements IEmployee.OffsetBalance
        Public Property AgencyID As Integer? Implements IEmployee.AgencyID
        Public Property Image As Byte() Implements IEmployee.Image
        Public Property AdvancementPoints As Integer Implements IEmployee.AdvancementPoints
        Public Property BPIInsurance As Decimal Implements IEmployee.BPIInsurance
        Public Property BranchID As Integer?

        <ForeignKey("PositionID")>
        Public Overridable Property Position As Position

        <ForeignKey("PayFrequencyID")>
        Public Overridable Property PayFrequency As PayFrequency

        Public Overridable Property Salaries As ICollection(Of Salary)

        Public ReadOnly Property MiddleInitial As String
            Get
                Return If(String.IsNullOrEmpty(MiddleName), Nothing, MiddleName.Substring(0, 1))
            End Get
        End Property

        Public ReadOnly Property IsDaily As Boolean
            Get
                Return (EmployeeType.ToLower = "daily") '"Daily"
            End Get
        End Property

        Public ReadOnly Property IsMonthly As Boolean
            Get
                Return (EmployeeType.ToLower = "monthly") '"Monthly"
            End Get
        End Property

        Public ReadOnly Property IsFixed As Boolean
            Get
                Return (EmployeeType.ToLower = "fixed") '"Fixed"
            End Get
        End Property

        Public ReadOnly Property IsWeeklyPaid As Boolean
            Get
                Return PayFrequencyID.Value = PayFrequencyType.Weekly
            End Get
        End Property

        Public ReadOnly Property IsPremiumInclusive As Boolean
            Get
                Return IsMonthly OrElse IsFixed
            End Get
        End Property

        Public ReadOnly Property FullNameLastNameFirst As String
            Get
                Return $"{LastName}, {FirstName}"
            End Get
        End Property

        Public ReadOnly Property FullNameWithMiddleInitialLastNameFirst As String
            Get
                Return $"{LastName}, {FirstName} {MiddleInitial}"
            End Get
        End Property

        Public ReadOnly Property FullNameWithMiddleInitial As String
            Get
                Return $"{FirstName} {If(MiddleInitial Is Nothing, "", MiddleInitial & ". ")}{LastName}"
            End Get
        End Property

        Public ReadOnly Property FullName As String
            Get
                Return $"{FirstName} {LastName}"
            End Get
        End Property

        Public ReadOnly Property EmployeeIdWithPositionAndEmployeeType As String
            Get
                Return $"ID# {EmployeeNo}, {Position?.Name}, {EmployeeType} Salary"
            End Get
        End Property

        Public ReadOnly Property IsUnderAgency As Boolean
            Get
                Return AgencyID.HasValue
            End Get
        End Property

        Public ReadOnly Property SssSchedule As String
            Get
                Return If(
                    IsUnderAgency,
                    Position?.Division?.AgencySssDeductionSchedule,
                    Position?.Division?.SssDeductionSchedule)
            End Get
        End Property

        Public ReadOnly Property PhilHealthSchedule As String
            Get
                Return If(
                    IsUnderAgency,
                    Position?.Division?.AgencyPhilHealthDeductionSchedule,
                    Position?.Division?.PhilHealthDeductionSchedule)
            End Get
        End Property

        Public ReadOnly Property PagIBIGSchedule As String
            Get
                Return If(
                    IsUnderAgency,
                    Position?.Division?.AgencyPagIBIGDeductionSchedule,
                    Position?.Division?.PagIBIGDeductionSchedule)
            End Get
        End Property

        Public ReadOnly Property WithholdingTaxSchedule As String
            Get
                Return If(
                    IsUnderAgency,
                    Position?.Division?.AgencyWithholdingTaxSchedule,
                    Position?.Division?.WithholdingTaxSchedule)
            End Get
        End Property

        Public ReadOnly Property IsActive As Boolean
            Get
                Return IsResigned = False AndAlso IsTerminated = False AndAlso IsRetired = False
            End Get
        End Property

        Public ReadOnly Property IsResigned As Boolean
            Get
                Return EmploymentStatus.Trim().ToUpper() = "RESIGNED"
            End Get
        End Property

        Public ReadOnly Property IsTerminated As Boolean
            Get
                Return EmploymentStatus.Trim().ToUpper() = "TERMINATED"
            End Get
        End Property

        Public ReadOnly Property IsRetired As Boolean
            Get
                Return EmploymentStatus.ToUpper().Trim() = "RETIRED"
            End Get
        End Property

    End Class

End Namespace

File: Accupay.Data/Entities/Employee.cs
using AccuPay.Data.Enums;
using AccuPay.Data.Helpers;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;

namespace AccuPay.Data.Entities
{
    [Table("employee")]
    public class Employee : BaseOrganizationalEntity
    {
        public int? PositionID { get; set; }
        public int? PayFrequencyID { get; set; }
        public string Salutation { get; set; }
        public string FirstName { get; set; }
        public string MiddleName { get; set; }
        public string LastName { get; set; }

        [Column("EmployeeID")]
        public string EmployeeNo { get; set; }

        public string TinNo { get; set; }
        public string SssNo { get; set; }
        public string HdmfNo { get; set; }
        public string PhilHealthNo { get; set; }
        public string EmploymentStatus { get; set; }
        public string EmailAddress { get; set; }
        public string WorkPhone { get; set; }
        public string HomePhone { get; set; }
        public string MobilePhone { get; set; }
        public string HomeAddress { get; set; }
        public string Nickname { get; set; }
        public string Gender { get; set; }
        public string EmployeeType { get; set; }
        public string MaritalStatus { get; set; }
        public DateTime BirthDate { get; set; }
        public DateTime StartDate { get; set; }
        public DateTime? TerminationDate { get; set; }
        public int? NoOfDependents { get; set; }

        //TODO: delete this
        public decimal LeaveBalance { get; set; }

        //TODO: delete this
        public decimal SickLeaveBalance { get; set; }

        [Column("LeaveAllowance")]
        public decimal VacationLeaveAllowance { get; set; }

        public decimal SickLeaveAllowance { get; set; }
        public decimal MaternityLeaveAllowance { get; set; }
        public decimal OtherLeaveAllowance { get; set; }
        public bool AlphalistExempted { get; set; }

        [Obsolete("Moved to employment policy")]
        public decimal WorkDaysPerYear { get; set; }

        public int? DayOfRest { get; set; }
        public string AtmNo { get; set; }
        public string BankName { get; set; }

        [Obsolete("Moved to employment policy")]
        public bool CalcHoliday { get; set; }

        [Obsolete("Moved to employment policy")]
        public bool CalcSpecialHoliday { get; set; }

        [Obsolete("Moved to employment policy")]
        public bool CalcNightDiff { get; set; }

        [Obsolete("Moved to employment policy")]
        public bool CalcRestDay { get; set; }

        public DateTime? DateRegularized { get; set; }
        public DateTime? DateEvaluated { get; set; }
        public bool RevealInPayroll { get; set; }

        [Obsolete("Moved to employment policy")]
        public decimal LateGracePeriod { get; set; }

        public decimal OffsetBalance { get; set; }
        public int? AgencyID { get; set; }
        public Byte[] Image { get; set; }
        public int AdvancementPoints { get; set; }
        public decimal BPIInsurance { get; set; }
        public int? BranchID { get; set; }
        public int? EmploymentPolicyId { get; set; }

        [ForeignKey("PositionID")]
        public virtual Position Position { get; set; }

        [ForeignKey("BranchID")]
        public virtual Branch Branch { get; set; }

        [ForeignKey("AgencyID")]
        public virtual Agency Agency { get; set; }

        [ForeignKey("PayFrequencyID")]
        public virtual PayFrequency PayFrequency { get; set; }

        public virtual EmploymentPolicy EmploymentPolicy { get; set; }

        public virtual ICollection<TimeEntry> TimeEntries { get; set; }

        public virtual ICollection<Paystub> Paystubs { get; set; }

        public int? OriginalImageId { get; set; }

        [ForeignKey("OriginalImageId")]
        public virtual File OriginalImage { get; set; }

        public string MiddleInitial
            => string.IsNullOrEmpty(MiddleName) ? null : MiddleName.Substring(0, 1);

        public bool IsDaily => EmployeeType?.ToLower() == "daily";

        public bool IsMonthly => EmployeeType?.ToLower() == "monthly";

        public bool IsFixed => EmployeeType?.ToLower() == "fixed";

        public bool IsWeeklyPaid => PayFrequencyID == (int)PayFrequencyType.Weekly;

        public bool IsPremiumInclusive => IsMonthly || IsFixed;

        public bool IsUnderAgency => AgencyID.HasValue;

        public string FullName => $"{FirstName} {LastName}".Trim();

        public string FullNameLastNameFirst => $"{LastName}, {FirstName}".Trim();

        public string FullNameWithMiddleInitialLastNameFirst =>
            $"{LastName}, {FirstName} {(MiddleInitial == null ? "" : MiddleInitial + ".")}".Trim();

        public string FullNameWithMiddleInitial =>
            $"{FirstName} {(MiddleInitial == null ? "" : MiddleInitial + ". ")}{LastName}".Trim();

        public string EmployeeIdWithPositionAndEmployeeType
            => $"ID# {EmployeeNo}, {Position?.Name}, {EmployeeType} Salary";

        public string SssSchedule =>
            IsUnderAgency ? Position?.Division?.AgencySssDeductionSchedule :
                            Position?.Division?.SssDeductionSchedule;

        public string PhilHealthSchedule => IsUnderAgency ?
                        Position?.Division?.AgencyPhilHealthDeductionSchedule :
                        Position?.Division?.PhilHealthDeductionSchedule;

        public string PagIBIGSchedule => IsUnderAgency ?
                        Position?.Division?.AgencyPagIBIGDeductionSchedule :
                        Position?.Division?.PagIBIGDeductionSchedule;

        public string WithholdingTaxSchedule => IsUnderAgency ?
                        Position?.Division?.AgencyWithholdingTaxSchedule :
                        Position?.Division?.WithholdingTaxSchedule;

        public bool IsActive => IsResigned == false && IsTerminated == false && IsRetired == false;

        public bool IsResigned => EmploymentStatus.Trim().ToUpper() == "RESIGNED";

        public bool IsTerminated => EmploymentStatus.Trim().ToUpper() == "TERMINATED";

        public bool IsRetired => EmploymentStatus.Trim().ToUpper() == "RETIRED";

        public bool IsFirstPay(PayPeriod payPeriod) =>
            payPeriod != null &&
            payPeriod.PayFromDate.Date <= StartDate.Date &&
            StartDate.Date <= payPeriod.PayToDate.Date;

        public static Employee NewEmployee(int organizationId, int userId)
        {
            return new Employee
            {
                RowID = null,
                OrganizationID = organizationId,
                Created = DateTime.Now,
                CreatedBy = userId,
                PayFrequencyID = PayrollTools.PayFrequencySemiMonthlyId,
                CalcHoliday = true,
                CalcNightDiff = true,
                CalcRestDay = true,
                CalcSpecialHoliday = true
            };
        }
    }
}
